### ISSUE TITLE
Expand specified GOODBYE reasons

### ISSUE THREAD

**Issue by theduke** [wamp-proto/wamp-proto#242]
In the context of webapps, I often have the need to reconnect in order to authenticate with new credentials. For this use case, a router closing the transport is cumbersome, since opening a new transport swallows up time.

On the other hand, it's not great for performance to have to router always wait for a new hello, since most GOODBYEs will result in a closed transport. 

It woud be good to have a standardized way for a GOODBYE message to signify the intent to reconnect in the details. A router could then refrain to close the transport immediately and wait for a new HELLO for some time.

Proposal for details:

```javascript
{
  "expectReconnect": true
  // alternative:
  "expectHello": true
}
```

Not sure if this is really useful, since clients would usually just close the transport anyway.,
What do you think?

**Comment by oberstet** [comment]
Just a quick note: AutobahnJS, upon receiving a GOODBYE will _always_ close the underlying transport anyway. AutobahnPython (when used as a WAMP client) has the ability to start a new session without closing the (existing) transport connection.

**Comment by theduke** [comment]
I'm currently working on a high performance C++ router implementation, that's why I'm thinking about it.

**Comment by oberstet** [comment]
I guess we have that already: GOODBYE has a `reason` (an URI) exactly for that purpose: https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/basic/bp_messages.md#goodbye

**Comment by theduke** [comment]
Agreed, the reason is probably ideal, but that does not help with a implementation agnostic solution.

Standardize a URI for reconnecting?
`wamp.error.reconnecting` or something like that?

**Comment by oberstet** [comment]
The currently defined goodbye reasons are here: https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/basic/bp_uris.md#session-close

Rdg `wamp.error.reconnecting`: that doesn't tell much. What should that mean? Also: whether a router will close the underlying transport upon a client initated GOODBYE shouldn't be something under client control.

**Comment by oberstet** [comment]
I have retitled and reopened: I agree in general, we might want to have more and more differentiated GOODBYE reasons. If you are interested in this, a systematic review / collection / brainstorming would be welcome!

**Comment by oberstet** [comment]
As far as I understand your itch (saving time when re-authenticating by not closing the underlying transport and do a full reconnect): I think it would be better to hint at the **intent** the client has:

* `wamp.error.reauthenticating` or similar

Because then, the router can take an informed decision. It might ignore it anyway. It might keep the transport open - but only for N seconds. Etc.

Because a client might want to re**connect** for other reasons too.

**Comment by theduke** [comment]
Agreed to this not being in anyway binding for the router.
That's why I wrote `A router COULD refrain`.

Just a hint for the router that it should expect a new HELLO and keep the transport open.
Of course, the client would have to check if the transport actually remains open after the routers GOODBYE ack and open a new transport if required.

A though regarding the GOODBYE reasons, that I already had before this:
`wamp.error.goodbye_and_out` seems kind of odd to me, since it is not actually any kind of error, just a reason. Just like `wamp.error.reauthenticating` is not an error.

Maybe another prefix would be more suitable?

* `wamp.status.goodbye_and_out`
* `wamp.status.reauthenticating`

**Comment by angiolep** [comment]
My understanding of transport and sessions is that they are separate entities operating at different levels. The former carries messages between peers to open/close one or more of the latter. I like the verbs _connect_/_disconnect_ when referring to the transport (which could be WebSocket or TCP), and prefer the verbs _open_/_close_ when referring to the sessions. Since this discussion is about WAMP Basic Profile, the authentication which theduke refers to is probably meant at the transport level, for example using TLS credentials (please confirm).

The GOODBYE message has clearly been designed to **close the current session** (if exists) and **it does NOT imply the peers will have to disconnect the underlying transport** upon sending or receiving it. I'm amused that some routers have been implemented to automatically disconnect the transport upon receiving the GOODBYE message as I believe that should be kept alive instead. My [akka-wamp](https://github.com/angiolep/akka-wamp) implementation (for example) does NOT disconnect the transport until a proper disconnect request is sent at the transport level (for example the FIN request at the TCP level, or the Close frame with opcode of 0x8 at the WebSocket level). Connecting a new transport takes usually longer than opening a session (that's why theduke would keep it alive, somehow), so I thought it would have been better to do nothing more than closing the current session on GOODBYE  (especially because the actual WAMP specification doesn't tell what to do with the transport).

I would not introduce any additional goodbye reasons to the WAMP protocol specification but since I'm now realizing that routers which automatically disconnect the transport have already been implemented by others, then I would add something like these:

``wamp.goodbye.keep_alive``
``wamp.goodbye.disconnect``

as optional reasons the client could send. Then the router should not ignore those but rather act as requested or just do what it prefers to do if the client doesn't ask for anything about the transport.

Like theduke, I also don't like the ``wamp.error`` prefix in the GOODBYE reasons as the word _error_ makes me think to something bad.

**Comment by goeddea** [comment]
Agree regarding the separation of transports and sessions (though we ourselves couple these e.g. in AutobahnJS).

**Comment by oberstet** [comment]
Whatever we do, I think we should nail it now;) will comment on https://github.com/wamp-proto/wamp-proto/issues/192 with my view ..

**Comment by gammazero** [comment]
@angiolep I agree with everything you said, up to and including:
> I would not introduce any additional goodbye reasons to the WAMP protocol specification

I do not think WAMP should add a `keep_alive` or `disconnect` request as part of the `GOODBYE`.  That is explicitly mixing WAMP with transport behavior.  What WAMP currently specifies is sufficient for both router implementations: those that leave transport open (to be closed at client's discretion) and those that close connection upon receipt of `GOODBYE`.

The spec states [here](http://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html#session-closing):

`A WAMP session ... ends when the underlying transport disappears or when the WAMP session is closed explicitly by a GOODBYE message sent by one Peer and a GOODBYE message sent from the other Peer in response.`

So, both router implementations are correct.  For routers that do not care to preserve the transport, echoing the client's `GOODBYE` is unnecessary and only a formality. Closing the transport is also an acceptable way to end the session.  I think the spec should clarify that,

`A peer sending a GOODBYE should expect to receive a GOODBYE from the other Peer in response, or for the transport to be closed in response.  A GOODBYE response is only required when the Peer receiving the first GOODBYE chooses to leave the transport open.`

Note the use of the word "may".  It is still OK for a peer to send a `GOODBYE` response and then also close the transport.  I think that covers all variations of router implementations.

This section on [GOODBYE](http://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html#goodbye) that states,

`Must be echo'ed by the receiving Peer.`

 should be amended to:

`Should be echo'ed by the receiving Peer, and must be echo'ed if the receiving peer does not close the transport.`

In other words, even if closing the transport, it is still polite to echo the `GOODBYE` as that allows Peers to differentiate between a graceful session close or the router dying.

**Comment by oberstet** [comment]
Ok, then why don't we also allow a client to just close a transport instead of sending a GOODBYE? I mean, if the router is allowed to be "rude" (closing transport, instead of echo'ing GOODBYE), then why not the client? I mean, that would then cover all 4 cases:

* client closes transport, and hence the router looses its transport automatically
* client sends GOODBYE, router closes transport
* client sends GOODBYE, router sends GOODBYE and closes transport
* client sends GOODBYE, router sends GOODBYE

**Comment by oberstet** [comment]
Btw, closely rereading the whole thread, there are actually two things here IMO:

* the necessity of echoing GOODBYE at all
* and what @theduke comments here https://github.com/wamp-proto/wamp-proto/issues/242#issuecomment-214216005: two news GOODBYE reasons:

* `wamp.status.goodbye_and_out`: the router hinting to the client that it will close the transport anyways (no chance for opening a new WAMP session on the same transport)
* `wamp.status.reauthenticating`: client hinting to the router that it indends to start a new WAMP session on the same transport without closing. "keep alive".

**Comment by KSDaemon** [comment]
Some my thoughts on topic:
* I also think, that replying with `wamp.close.goodbye_and_out` for normal connection is unnecessary. So i propose to remove it from spec explicitly.
* Regarding `wamp.status.reauthenticating` its not clear what to do:
   * Client peer received `wamp.status.reauthenticating` from router. What's next? Do client peer need to automatically send HELLO message, like it is first time. Or client peer need to wait for some other message from router? What message? What status of client peer is for this period? I mean how it looks like for a developer, who uses client implementation? Is it still connected? Or is it closed? Do client implementation need to fire some possible callbacks (like ondisconnect, onclose, etc)?
   * Can be such a case, when clients sends `wamp.status.reauthenticating`? What's next in this case? Do router need to cleanup session state (subscriptions, registered rpcs, ongoing calls, etc)? Or simply close session? In this case whats the difference with `wamp.close.goodbye_and_out`? Espessially if the initiator peer will send this message and disconnect underlying transport. 
   * I don't think that router should disconnect transport upon session close, initiated by client peer. In most cases, client peer will send GOODBYE and disconnect. If client peer decided to reuse transport for new session — it simple send HELLO message. So, imo, when router receives GOODBYE message — it should just close WAMP session, cleanup registrations/subscriptions and that is all. After that router can receive again new HELLO message or client disconnect.
